Source/Tests/LinuxSyscalls/Syscalls.h

```diff
@@ -206,6 +218,97 @@ class SyscallHandler : public FEXCore::HLE::SyscallHandler {
   #endif
 
   std::unique_ptr<FEX::HLE::MemAllocator> Alloc32Handler{};
+  
+  ///// VMA (Virtual Memory Area) tracking /////
+
+  FEXCore::HLE::AOTIRCacheEntryLookupResult LookupAOTIRCacheEntry(uint64_t GuestAddr) final override;
+
+
+  struct MRID {
+    uint64_t dev; /* kernel only uses lower 32-bits of dev_t */
+    uint64_t id;
+
+    bool operator<(const MRID& other) const {
+      return std::tie(dev, id) < std::tie(other.dev, other.id);
+    }
+  };
+
+  // We use bits 63:32 for our own custom dev tracking
+  static constexpr const uint64_t MRID_PRIVATE = 0x1'0000'0000ULL;
+  static constexpr const uint64_t MRID_ANON = 0x2'0000'0000ULL;
+  static constexpr const uint64_t MRID_SHM =  0x3'0000'0000ULL;
+
+  struct VMAEntry;
+
+  struct MappedResource {
+    typedef std::map<MRID, MappedResource> ContainerType;
+
+    FEXCore::IR::AOTIRCacheEntry *AOTIRCacheEntry;
+    VMAEntry *FirstVMA;
+    uint64_t Length; // 0 if not fixed size
+    ContainerType::iterator Iterator;
+  };
+
+  struct VMAProt {
+    bool Writable: 1;
+    bool operator==(const VMAProt& other) const {
+      return other.Writable == Writable;
+    }
+  };
+
+  struct VMAOptions {
+    VMAOptions(int Writable, const int Shared) : Prot {Writable != 0}, Flags {Shared != 0} { }
+
+    VMAProt Prot;
+
+    const struct {
+      bool Shared: 1;
+    } Flags;
+  };
+
+  struct VMAEntry {
+    MappedResource* Resource;
+    VMAEntry* ResourcePrevVMA;
+    VMAEntry* ResourceNextVMA;
+
+    uint64_t Base;
+    uint64_t Offset;
+    uint64_t Length;
+
+    VMAOptions Options;
+  };
+
+  struct VMATracking {
+    using VMAEntry = SyscallHandler::VMAEntry;
+    std::shared_mutex Mutex{};
+    
+    // Memory ranges indexed by page aligned starting address
+    typedef std::map<uint64_t, VMAEntry> VMAsType;
+    VMAsType VMAs;
+    MappedResource::ContainerType MappedResources;
+
+    // Mutex must be at least shared_locked before calling
+    VMAsType::const_iterator LookupVMAUnsafe(uint64_t GuestAddr) const;
+
+    // Mutex must be unique_locked before calling
+    void SetUnsafe(FEXCore::Context::Context *Ctx, MappedResource *MappedResource, uintptr_t Base, uintptr_t Offset, uintptr_t Length, VMAOptions Flags);
+    
+    // Mutex must be unique_locked before calling
+    void ClearUnsafe(FEXCore::Context::Context *Ctx, uintptr_t Base, uintptr_t Length, MappedResource *PreservedMappedResource = nullptr);
+
+    // Mutex must be unique_locked before calling
+    void ChangeUnsafe(uintptr_t Base, uintptr_t Length, VMAProt Prot);
+
+    // Mutex must be unique_locked before calling
+    // Returns the Size fo the Shm or 0 if not found
+    uintptr_t ClearShmUnsafe(FEXCore::Context::Context *Ctx, uintptr_t Base);
+  private:
+    bool ListRemove(VMAEntry *Mapping);
+    void ListReplace(VMAEntry *Mapping, VMAEntry *NewMapping);
+    void ListInsert(VMAEntry *AfterMapping, VMAEntry *NewMapping);
+    void ListPrepend(MappedResource *Resource, VMAEntry *NewVMA);
```

Comment:
Because it needs the `MappedResource` to prepend to it. If the list is empty, there's no `VMAEntry` to pass.

Reply:
Maybe this is another candidate that could be cleaned up with an explicit `struct VMAEntriesList` type?

Reply:
It would still need a `MappedResource`. That's one of the benefits of not using std::list, that the pointers are intrusive and we don't need to resolve the list for many operations.

Reply:
Maybe this is another candidate that could be cleaned up with an explicit `struct VMAEntriesList` type?

Something like this should do:
```cpp
struct VMAEntriesList {
    VMAEntry *head; // or nullptr if empty
};

struct MappedResource {
    VMAEntriesList FirstVMA;
};

void ListPrepend(VMAEntriesList*, VMAEntry*);

ListPrepend(MappedResource->FirstVMA, &VMAInserted.first->second);
```

Reply:
Sorry, I intended to replace my previous one-liner comment with the extended one below, but then messed up.

Does your comment about `ListPrepend` still needing a MappedResource still apply with this concrete proposal?

Reply:
Then it would work - though I'd go for a `VMAEntriesList *` there, not &. Tbh, I don't see much difference between as is and this change, though

Reply:
Put a bit more though in this, and as the system will be re-written to use some O(logn) container, and it only ever has one use in single file private to syscalls, I don't want to spend the engineering time for this change.